comtypes/npsupport.py

```diff
@@ -1,125 +1,158 @@
+            '<u8', '<u4', '<f4', '<f8', '<i8', ptr_typecode, ptr_typecode,
+            ptr_typecode, ptr_typecode, _tagBRECORD_format,
+        ],
+        offsets=[0] * 19  # This is what makes it a union
+    )
+
+    tagVARIANT_format = [
+        ("vt", '<u2'),
+        ("wReserved1", '<u2'),
+        ("wReserved2", '<u2'),
+        ("wReserved3", '<u2'),
+        ("_", U_VARIANT_format),
+    ]
+
+    return numpy.dtype(tagVARIANT_format)
+
+
+def isndarray(value):
+    """ Check if a value is an ndarray.
+
+    This cannot succeed if numpy is not available.
+
+    """
+    if not HAVE_NUMPY:
+        if hasattr(value, "__array_interface__"):
+            raise ValueError(
+                (
+                    "Argument {0} appears to be a numpy.ndarray, but "
+                    "comtypes numpy support has not been enabled. Please "
+                    "try calling comtypes.npsupport.enable_numpy_interop() "
+                    "before passing ndarrays as parameters."
+                ).format(value)
+            )
+        return False
+    numpy = get_numpy()
+    return isinstance(value, numpy.ndarray)
+
+
+def isdatetime64(value):
+    """ Check if a value is a datetime64.
+
+    This cannot succeed if datetime64 is not available.
+
+    """
+    if not HAVE_NUMPY:
+        return False
+    return isinstance(value, datetime64)
+
+
+def _check_ctypeslib_typecodes():
+    import numpy as np
+    from numpy import ctypeslib
+    try:
+        from numpy.ctypeslib import _typecodes
+    except ImportError:
+        from numpy.ctypeslib import as_ctypes_type
+
+        dtypes_to_ctypes = {}
+
+        for tp in set(np.sctypeDict.values()):
+            try:
+                ctype_for = as_ctypes_type(tp)
+                dtypes_to_ctypes[np.dtype(tp).str] = ctype_for
+            except NotImplementedError:
+                continue
+        ctypeslib._typecodes = dtypes_to_ctypes
+    return ctypeslib._typecodes
+
+
+def enable_numpy_interop():
+    """ Import the numpy library (if not already imported) and set up the
+    necessary functions for comtypes to work with ndarrays.
+
+    """
+    global HAVE_NUMPY
+    if HAVE_NUMPY:
+        return
+
+    global numpy
+    import numpy
+
+    HAVE_NUMPY = True
+    global typecodes
+    typecodes = _check_ctypeslib_typecodes()
+    # dtype for VARIANT. This allows for packing of variants into an array, and
+    # subsequent conversion to a multi-dimensional safearray.
+    global VARIANT_dtype
+    try:
+        VARIANT_dtype = _make_variant_dtype()
+    except ValueError:
+        pass
+    global datetime64
+    global com_null_date64
+    # This simplifies dependent modules
+    try:
+        from numpy import datetime64
+    except ImportError:
+        pass
+    else:
+        try:
+            # This does not work on numpy 1.6
+            com_null_date64 = datetime64("1899-12-30T00:00:00", "ns")
+        except TypeError:
+            pass
```

Comment:
Too many silent `pass` statements in exception handlers is also not good.

Reply:
Well, this is legacy code that exists for a while. This comment can be ignored for now.